### DISCUSSION TITLE
Motion activated LED with timer to turn off

### DISCUSSION THREAD

**Discussion by musicman944** [HomeSpan/HomeSpan#238] [Q&A]
I want to have a motion sensor or occupancy sensor that will turn on a light.  From the examples and documentation, that seems rather straightforward.  However, I want the light to stay on for a period of time then turn off after all motion stops.  I have not yet found anything that seems to allow that type of operation.  Can anyone provide me with a hint or pointer on the appropriate way to do this?

**Comment by HomeSpan** [comment]
There are two approaches to this, depending on where you want to include the logic.  In principle, you could use HomeSpan to program a motion sensor only, and then have your Home App look for motion on that sensor and trigger another action, such as turning on a light.  The Home App could also be be able to turn the light off after a certain period of time, though this functionality keeps moving in different versions of iOS.  You could then use HomeSpan to create a separate device to control a light from the Home App, or you could use a commercial light-control product.

The other approach is to build everything into the ESP32 device.  In this case the motion sensor would be used in the ESP32 device, but you would not need to configure it as a HomeKit Service since the Home App would not need to know anything about the motion sensor. Instead you would use the motion sensor to trigger a light (also connected to the ESP32 device), keep track of time, and then turn the light off after no motion detected for some time.  The could all be programmed into an ESP32 device without using HomeSpan at all; where HomeSpan comes in is if you also want to control the light from your Home App, in which case you would configure the light in HomeSpan to connect to HomeKit.

**Reply by musicman944** [reply]
Thanks for the response and also for sharing HomeSpan!  My inclination is to write it all in the ESP32 because the light(s) will be primarily operated by the motion sensor.  There are occasions when I want to override the motion sensor and lock the light on.  My primary reason for using HomeKit is to provide the override function to bypass the motion sensor and explicitly control the light.  Would the motion sensor code go in the main loop and use standard Arduino code to turn the light on/off via digitalwrite?

**Comment by HomeSpan** [comment]
I would place the motion-sensor code/logic in the loop() method of a Homespan Lightbulb service instead of the main loop of the Arduino.  If embedded in the Lightbulb service you will be able to use the setVal() method to update the status of the light in HomeKit.  In this fashion, triggering the motion sensor will not only turn on the light, but will show that the light is ON inside the Home App on your iPhone (similarly, you can update the status to show when the light turns OFF).
